Common/Securities/Option/StrategyMatcher/OptionStrategyLegDefinition.cs

```diff
@@ -0,0 +1,208 @@
+            Multiplier = multiplier;
+        }
+    }
+
+    /// <summary>
+    /// Defines a single option leg in an option strategy. This definition supports direct
+    /// match (does position X match the definition) and position collection filtering (filter
+    /// collection to include matches)
+    /// </summary>
+    public class OptionStrategyLegDefinition : IEnumerable<OptionStrategyLegPredicate>
+    {
+        /// <summary>
+        /// Gets the unit quantity
+        /// </summary>
+        public int Quantity { get; }
+
+        /// <summary>
+        /// Gets the contract right
+        /// </summary>
+        public OptionRight Right { get; }
+
+        private readonly OptionStrategyLegPredicate[] _predicates;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="OptionStrategyLegDefinition"/> class
+        /// </summary>
+        /// <param name="right">The leg's contract right</param>
+        /// <param name="quantity">The leg's unit quantity</param>
+        /// <param name="predicates">The conditions a position must meet in order to match this definition</param>
+        public OptionStrategyLegDefinition(OptionRight right, int quantity, IEnumerable<OptionStrategyLegPredicate> predicates)
+        {
+            Right = right;
+            Quantity = quantity;
+            _predicates = predicates.ToArray();
+        }
+
+        /// <summary>
+        /// Yields all possible matches for this leg definition held within the collection of <paramref name="positions"/>
+        /// </summary>
+        /// <param name="legs">The preceding legs already matched for the parent strategy definition</param>
+        /// <param name="positions">The remaining, unmatched positions available to be matched against</param>
+        /// <returns>An enumerable of potential matches</returns>
+        public IEnumerable<OptionStrategyLegDefinitionMatch> Match(
+            IReadOnlyList<OptionPosition> legs,
+            OptionPositionCollection positions
+            )
+        {
+            foreach (var position in Filter(legs, positions))
+            {
+                var multiplier = position.Quantity / Quantity;
+                if (multiplier != 0)
+                {
+                    yield return new OptionStrategyLegDefinitionMatch(multiplier,
+                        position.WithQuantity(multiplier * Quantity)
+                    );
+                }
+            }
+        }
+
+        /// <summary>
+        /// Filters the provided <paramref name="positions"/> collection such that any remaining positions are all
+        /// valid options that match this leg definition instance.
+        /// </summary>
+        public OptionPositionCollection Filter(IReadOnlyList<OptionPosition> legs, OptionPositionCollection positions)
+        {
+            foreach (var predicate in _predicates)
+            {
+                positions = predicate.Filter(legs, positions, true);
```

Comment:
Shouldn't we only filter if all options predicate legs are respected? 🤔

Reply:
This is how we determine if all predicates are respected. We filter the collection down using the predicates and if anything is remaining then they all matched. Now, what I can add here is `if (positions.IsEmpty) break;` -- no sense is filtering an empty collection.